ArjunaJTA/jms/src/main/java/org/jboss/narayana/jta/jms/classic/ConnectionProxy.java

```diff
@@ -0,0 +1,158 @@
+
+    private final TransactionHelper transactionHelper;
+
+    public ConnectionProxy(XAConnection xaConnection,
+            TransactionHelper transactionHelper) {
+
+        this.xaConnection = xaConnection;
+        this.transactionHelper = transactionHelper;
+    }
+
+    @Override
+    public Session createSession(boolean transacted, int acknowledgeMode)
+            throws JMSException {
+
+        if (transactionHelper.isTransactionAvailable()) {
+            return createAndEnlistSession();
+        }
+
+        return xaConnection.createSession(transacted, acknowledgeMode);
+    }
+
+    @Override
+    public void close() throws JMSException {
+        if (transactionHelper.isTransactionAvailable()) {
+            Synchronization synchronization =
+                    new ConnectionClosingSynchronization(xaConnection);
+            transactionHelper.registerSynchronization(synchronization);
+
+            if (LOGGER.isTraceEnabled()) {
+                LOGGER.trace("Registered synchronization to close the connection: "
+                        + synchronization);
+            }
+        } else {
+            xaConnection.close();
+        }
+    }
+
+    @Override
+    public String getClientID() throws JMSException {
+        return xaConnection.getClientID();
+    }
+
+    @Override
+    public void setClientID(String clientID) throws JMSException {
+        xaConnection.setClientID(clientID);
+    }
+
+    @Override
+    public ConnectionMetaData getMetaData() throws JMSException {
+        return xaConnection.getMetaData();
+    }
+
+    @Override
+    public ExceptionListener getExceptionListener() throws JMSException {
+        return xaConnection.getExceptionListener();
+    }
+
+    @Override
+    public void setExceptionListener(ExceptionListener listener) throws JMSException {
+        xaConnection.setExceptionListener(listener);
+    }
+
+    @Override
+    public void start() throws JMSException {
+        xaConnection.start();
+    }
+
+    @Override
+    public void stop() throws JMSException {
+        xaConnection.stop();
+    }
+
+    @Override
+    public ConnectionConsumer createConnectionConsumer(Destination destination,
+            String messageSelector, ServerSessionPool sessionPool,
+            int maxMessages) throws JMSException {
+
+        return xaConnection.createConnectionConsumer(destination,
+                messageSelector, sessionPool, maxMessages);
+    }
+
+    @Override
+    public ConnectionConsumer createDurableConnectionConsumer(Topic topic,
+            String subscriptionName, String messageSelector,
+            ServerSessionPool sessionPool, int maxMessages) throws JMSException {
+
+        return xaConnection.createDurableConnectionConsumer(topic,
+                subscriptionName, messageSelector, sessionPool, maxMessages);
+    }
+
+    private Session createAndEnlistSession() throws JMSException {
+        if (!transactionHelper.isTransactionActive()) {
```

Comment:
I think this is not right here. It is totally possible that the transaction can go inactive immediately after so best to just leave the exception up to enlistResource.